src/python/pants/core/util_rules/strip_source_roots.py

```diff
@@ -93,20 +96,26 @@ def determine_source_root(path: str) -> str:
         return PurePath(path).parent.as_posix()
 
     if request.representative_path is not None:
-        resulting_digest = await Get[Digest](
-            RemovePrefix(
-                request.snapshot.digest, determine_source_root(request.representative_path),
-            )
-        )
-        resulting_snapshot = await Get[Snapshot](Digest, resulting_digest)
-        return SourceRootStrippedSources(snapshot=resulting_snapshot)
+        source_root = determine_source_root(request.representative_path)
+        is_repo_level_source_root = source_root == ""
+        if is_repo_level_source_root:
+            return SourceRootStrippedSources(request.snapshot)
+        resulting_snapshot = await Get[Snapshot](RemovePrefix(request.snapshot.digest, source_root))
+        return SourceRootStrippedSources(resulting_snapshot)
 
     files_grouped_by_source_root = {
         source_root: tuple(files)
         for source_root, files in itertools.groupby(
             request.snapshot.files, key=determine_source_root
         )
     }
+
+    only_has_repo_level_source_root = (
+        len(files_grouped_by_source_root) == 1 and next(iter(files_grouped_by_source_root)) == ""
```

Comment:
What do you think about fixing the intrinsic(s)?:
https://github.com/pantsbuild/pants/blob/88d30dfebdcfc0a5efc55a61e67dec8cb5c4c757/src/rust/engine/src/intrinsics.rs#L167
https://github.com/pantsbuild/pants/blob/88d30dfebdcfc0a5efc55a61e67dec8cb5c4c757/src/rust/engine/src/intrinsics.rs#L182

Although this python code is the only user of the intrinsic today, even it needs to do the dance twice.

Reply:
I think it would be good to fix in Rust, but I also think we want these early returns in Python. 

In particular, we want it to be able to avoid calling all 3 of `await MultiGet[Snapshot](SnapshotSubset)`, `await MultiGet[Digest](RemovePrefix)`, and `await Get[Digest](MergeDigests)`. No matter how many optimizations we make to Rust, I believe there will always be a performance hit to making those calls.

--

Still, I'll try making the change and see the impact. (Benjy taught me the importance of benchmarks with performance work, rather than intuition.)

Reply:
Upon thinking more about this, I think this optimization belongs in Python, rather than Rust, for two reasons:

1) Benjy's point about what we show in `--v2-ui`. It's confusing to show the rule `remove_prefix` when you really aren't going to remove anything.
2) Most Python call sites already know what the prefix is, e.g. ~90% of call sites currently hardcoding it. No one would write `RemovePrefix(digest, "")`.